examples/sparse/sign.py

```diff
@@ -0,0 +1,111 @@
+"""
+[SIGN: Scalable Inception Graph Neural Networks]
+(https://arxiv.org/abs/2004.11198)
+
+This example shows a simplified version of SIGN: a precomputed 2-hops diffusion
+operators on top of symmetrically normalized adjacency matrix A.
+"""
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.optim import Adam
+
+from dgl.data import CoraGraphDataset
+from dgl.mock_sparse import create_from_coo, diag, identity
+
+################################################################################
+# (HIGHLIGHT) Take the advantage of DGL sparse API to significantly simplify the
+# code for sign diffusion operator.
+################################################################################
+def sign_diffusion(A, X, r):
+    # Perform the r-hops diffusion operation.
+    X_sign = [X]
+    for _ in range(r):
+        X = A @ X
+        X_sign.append(X)
+    return X_sign
+
+
+class SIGN(nn.Module):
+    def __init__(self, in_size, out_size, r, hidden_size=64):
+        super().__init__()
+        self.theta = nn.ModuleList(
+            [nn.Linear(in_size, hidden_size) for _ in range(r + 1)]
+        )
+        self.omega = nn.Linear(hidden_size * (r + 1), out_size)
+
+    def forward(self, X_sign):
+        results = []
+        for i in range(len(X_sign)):
+            results.append(self.theta[i](X_sign[i]))
+        Z = F.relu(torch.cat(results, dim=1))
+        return F.sigmoid(self.omega(Z))
+
+
+def evaluate(g, pred):
+    labels = g.ndata["label"]
+    val_mask = g.ndata["val_mask"]
+    test_mask = g.ndata["test_mask"]
+
+    # Compute accuracy on validation/test set.
+    val_acc = (pred[val_mask] == labels[val_mask]).float().mean()
+    test_acc = (pred[test_mask] == labels[test_mask]).float().mean()
+    return val_acc, test_acc
+
+
+def train(g, model):
+    labels = g.ndata["label"]
+    train_mask = g.ndata["train_mask"]
+    optimizer = Adam(model.parameters(), lr=3e-3)
+
+    for e in range(20):
+        # Forward.
+        logits = model(X_sign)
+
+        # Compute prediction.
+        pred = logits.argmax(1)
+
+        # Compute loss with nodes in training set.
+        loss = F.cross_entropy(logits[train_mask], labels[train_mask])
+
+        # Backward.
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+
+        # Evaluate the prediction.
+        val_acc, test_acc = evaluate(g, pred)
+        print(
+            f"In epoch {e}, loss: {loss:.3f}, val acc: {val_acc:.3f}, test acc"
+            f": {test_acc:.3f}"
```

Comment:
It's a bit controversial to print test accuracy every epoch as some people consider it as cheating.

Reply:
Since this is an example, it should be fine to show it.